Add facilities that allow for tracking of ancestral lines

ilyaZar opened the pull request:
So here are the discussed changes to allow for tracking of ancestral lines (just squashed my commits for the first time so the commit history will be cleaner than last time :) ). 

I agree with @adamjohansen about the documentation; we have the comments to the include files and the .Rd files for functions that are meant to be called via R or am I mistaken?

@eddelbuettel I upgraded `Rcpp` to the Drat version, so this time no changes to the export-files I think.

Hope the CI passes this time, I'll check this first thing in the morning!

LeahPrice commented on ChangeLog:
```diff
@@ -1,3 +1,8 @@
+2021-05-20  Ilya Zarubin  <ilya.a.zarub@gmail.com>
```
Perhaps we should update the date to avoid confusion when looking back at discussions.

ilyaZar replied:
Done, but I squashed my branch commits and `git push origin +historyelement-class-additions-for-AL ` hence the outdated label (also below)...

LeahPrice commented on ChangeLog:
```diff
@@ -1,3 +1,8 @@
+2021-05-20  Ilya Zarubin  <ilya.a.zarub@gmail.com>
+
+	* inst/include/history.h: add facilities for tracking of ancestral lines (mostly additions to the historyelement class)
```
We've been restricting the line lengths to I think around 80 characters/columns for readability, so these two dot points could be split across more lines. E.g. see your 2021-05-20 src/nonLinPMMH.cpp item.

Adam and Dirk use Emacs where this is all automatic.

LeahPrice reviewed (commented):
These changes look good to me, as discussed in #55, but I have a couple of formatting suggestions for the changelog.

ilyaZar commented:
@LeahPrice thanks for the remarks, all done!

adamjohansen commented on inst/include/sampler.h:
```diff
@@ -161,6 +162,15 @@ namespace smc {
         /// Returns the Effective Sample Size of the specified particle generation.
         double GetESS(long n) const {return History[n].GetESS();}
+        /// Returns a vector of population objects re-ordered to match ancestral lines of the particle system
+        std::vector<population<Space>> GetApop(void) const;
```
This is rather trivial, but should this be `GetAPop` or similar for consistency with the other use of A in method names around here?

ilyaZar replied:
Done, will keep in mind for next time!

adamjohansen commented on inst/include/sampler.h:
```diff
@@ -161,6 +162,15 @@ namespace smc {
+        // void GetALineInd(long n) const;
+        arma::Col<unsigned int> GetALineInd(long n) const;
+        std::vector<Space> GetALineSpace(long n) const;
+        /// HELPER FUNCTION THAT MAY BE DELETED; RETURNS uRSIndices
```
Are these functions intended to be used externally? If not should they be private or protected?

It just seems a bit unfriendly to out to the interface something with a "MAY BE DELETED" comment attached. I don't immediately see any harm in leaving these functions but if we do I'd be inclined to do that with the view to keeping them permanently (or as permanently as one does anything in code).

ilyaZar replied:
1. I think that depends on how we want to implement conditional SMC and grant the user access to these class members. 
* It would be more flexible to allow the user access to all trajectories, hence I'd suggest `GetALineSpace()` should be public.
* I believe that `GetALineSpace()` is only ever useful for internal purposes, specifically, also for derived cond. SMC. So `GetALineInd()` should be protected. 

If no objections I'll do as above.

2. Ah I see, my mistake, just wanted to highlight this with capslock for a discussion.
* I think, we might want to decide whether to make ancestral lines accessible only via the historyelement class (via `historyelement.SetAindices()`) or additionally (but this might be an unnecessary deflection since it only gives access to the current iteration resampled indices) via 
https://github.com/rcppsmc/rcppsmc/blob/6bd4ceaa976d9285cb1d867515bb003e0f3d7361/inst/include/sampler.h#L170-L171

* The same goes for 
https://github.com/rcppsmc/rcppsmc/blob/6bd4ceaa976d9285cb1d867515bb003e0f3d7361/inst/include/sampler.h#L172-L173

Maybe there are other merits to these functions, that I don't see. In general, I'd be a bit more inclined to delete unnecessary facilities (only keeping what's relevant, for clarity purposes). 

What do the others think @eddelbuettel @LeahPrice ?

adamjohansen replied:
I'll interrupt given I raised this in the first place.

Presumably in 1. the `GetALineSpace` in the second bullet point is a typo for `GetALineInd`. No objections from me, although I didn't want to insist on anything just that if we have functions which we might get rid of them exposing their interface might not be helpful. I still have no problem with keeping them; it's difficult to predict every possible use case and making everything available to users that might want it doesn't seem to hurt.

No strong views either way on 2. Largely a matter of taste, I think.

LeahPrice replied:
I also don't think we should make functions public unless we plan on keeping them, and I'm not too concerned about keeping functions that we can't see a direct use for right now if they might be useful later. I don't have any strong opinions on point 2.

adamjohansen commented on inst/include/sampler.h:
```diff
@@ -403,7 +413,71 @@ namespace smc {
+        /// 2. Storing these particle population classes in output: populationAL
+        if (historylength > 1){
+            for (int t = historylength - 2; t >= 0; --t){
+                for(int i = 0; i < static_cast<int>(N); ++i){
```
This looks like a bad idea.

If we use a long for `N` to allow for large particle sets on architectures with small words or similar then we should really be using a long for indices which iterate over the particle set shouldn't we?

If there's a reason to strong prefer int then there's probably no harm in changing the type of N from long to int given the architectures and compilers in current used. There aren't many 16 bit `int`s around any more.

eddelbuettel replied:
I also find `uint64_t and `int64_t` more obvious, I think C++ pros also like size_t for the indices. 

```r
> Rcpp::evalCpp("sizeof(long)")    
[1] 8 
> Rcpp::evalCpp("sizeof(int)") 
[1] 4   
> Rcpp::evalCpp("sizeof(int64_t)")      
[1] 8  
> Rcpp::evalCpp("sizeof(size_t)")    
[1] 8
> 
```

ilyaZar replied:
Changed!

adamjohansen commented on inst/include/sampler.h:
```diff
@@ -403,7 +413,71 @@ namespace smc {
+    {
+        long N = GetNumber();
+        long lHistorylength = GetHistoryLength();
+        int historylength = static_cast<int>(lHistorylength);
```
Is there a reason to keep casting long to int? In the abstract losing bits like this seems a bad idea, although `long` and `int` may not be different on some architectures.

adamjohansen commented on inst/include/sampler.h:
```diff
@@ -403,7 +413,71 @@ namespace smc {
+
+        /// Get final iteration indices and final iteration particle values
+        arma::Col<unsigned int> AlineInd(historylength);
+        AlineInd = this->GetALineInd(static_cast<int>(n));
```
Again there seems to be some inconsistency around case with these As. I don't want to obsess over this, but it does jar a bit when the function is `GetALineInd` and the variable is `AlineInd`.

ilyaZar replied:
Done, will keep in mind for next time!

adamjohansen commented on inst/include/sampler.h:
```diff
@@ -441,6 +515,8 @@ namespace smc {
         }
         else {
             nResampled = 0;
+            uRSIndices = arma::zeros<arma::Col<unsigned int> >(static_cast<int>(N));
```
I don't use armadillo enough, I suspect, but is there a reason to first create and initialise to zero and _then_ to fill with specified values? It naively seems a touch redundant.

eddelbuettel replied:
Yes, looking at 518 and 519 I think only 519 ("linspace") is needed.

Also, Arma has `typedef` for the common type so instead of arma::icovec and arma::ivec are the same as a arma::Col<unsiged int>.  I guess it takes use to make it more legible as it is new -- arma::Col<unsigned int> is understood by everybody but a mouthful.

ilyaZar replied:
1. Got rid of the unnecassry line.

2. On typedefs: agreed @eddelbuettel  , but I followed the existing code base like here:
* https://github.com/rcppsmc/rcppsmc/blob/6bd4ceaa976d9285cb1d867515bb003e0f3d7361/inst/include/sampler.h#L251-L255
* https://github.com/rcppsmc/rcppsmc/blob/6bd4ceaa976d9285cb1d867515bb003e0f3d7361/inst/include/sampler.h#L796

Maybe we should invest a bit of time and do a PR for re-naming everything to pure typedefs?

eddelbuettel replied:
Maybe.  We should pick a consistent style. I can see reasons for being explicit (everybody understands) and for being shorter (but maybe that is only used for more Armadillo-heavy projects?)

adamjohansen replied:
No strong views either way; but consistent is good.

adamjohansen commented on inst/include/sampler.h:
```diff
@@ -457,13 +533,21 @@ namespace smc {
+                break;
+            case HistoryType::AL:
+                histel.Set(N, pPopulation, nAccepted, nRepeats, historyflags(nResampled), uRSIndices);
+            /// To avoid compiler warnings, HistoryType::NONE is handled
```
Don't you want a `break` here? Rather than falling through to the NONE case?

ilyaZar replied:
Done!

adamjohansen commented on inst/include/sampler.h:
```diff
@@ -641,6 +725,8 @@ namespace smc {
         }
         else {
             nResampled = 0;
+            uRSIndices = arma::zeros<arma::Col<unsigned int> >(static_cast<int>(N));
```
I guess we could do this conditionally on wanting to store this information, but I can see the case for keeping things simple.

ilyaZar replied:
At first, an `if()` looked a bit ugly to me, so I deleted it for the moment. 

Now I believe, an external person might get confused by this redundancy in case of `HistoryType::RAM` so I added it now.

adamjohansen replied:
I don't have strong views, it just seems like an unnecessary operation if the information isn't wanted and I still remember when computers were slow...

adamjohansen reviewed (changes requested):
This looks good. There are a few details that I've commented on.

Aside from cases of bits in names (sorry, this is a pain but it seems like a good time to try to do something consistent here) the main thing that is bothering me is that we seem to do a lot of casting between `long` and `int` which although it will probably _just work_ for most people on most modern systems seems to me to be a bad thing in the abstract. This might reflect deficiencies in the existing code base that this PR was just trying to work around, but is it feasible to avoid needing to do it without undue effort or are there just too many moving parts?

(If it's necessary to have `int` rather than `long` in places for any reason then it would be arguably more sensible to replace the `long`s with `int`s than to mix the two.)

adamjohansen commented:
@ilyaZar we do have the documentation that you mention, but I've had it in mind that comments on the code together with man-page level documentation of some R functions which are essentially example implementations isn't really enough for someone to come to the code cold and implement their favourite algorithm with it.

Consequently, we ought when we have time and capacity to provide some more substantial documentation. We were, at one stage, trying to put together an analogue of [this SMCTC overview](http://www.jstatsoft.org/v30/i06/paper), with the aim of telling people roughly how the library works and how to us it along with a few examples but we never quite found the time to finish it.

At some point I think something like that is probably going to be necessary to make things accessible to the people who might find the code useful. It could just be a more detailed user manual than we currently have, or something like the SMCTC summary, but a fairly holistic description of how to go from nothing to a working SMC algorithm seems to be needed.

ilyaZar commented:
Allright. So before I do the my re-work of the PR:

1. Using typdef or explicit naming (above, I adhear to the existing explicit naming `arma::Col<unsigned int>` in the current codebase, but if we wish to be consistent then we probably ought to change it everywhere and not just for this particular PR?)
2.  @adamjohansen : the `long` vs. `int` is the same issue, I think it might be wise to decide now and do a PR fixing this (but not just for this commit).

I believe it's worth a "discussion" (or at least 1-2 seperate PR unrelated to this one because this one deals with the ancestral lines solely) because there are numerous incidents of 1./2. inconsistencies that we might want to address.

adamjohansen commented:
Thanks, @ilyaZar 
1. No strong views from me; the person writing the code tends to get to make these decisions unless someone has a reason to favour one or the other (@edelbuettel is definitely the armadillo expert...).
2. I was being lazy and hadn't checked whether we were doing this already and you'd been pushed into following what we were doing. If that's the case then it would probably make sense to resolve this PR without worrying about that and add an issue ticket to fix this throughout the code. If the existing code avoids this then we might as well keep things clean here.  A quick `grep static_cast *.h` does seem to reveal existing instances of this. It's not clear to me if this is because of an interaction with armadillo or if we've just been sloppy in the past, but it does seem that it's a wider issue than this PR and should probably be dealt with separately.
3. Examples, vignettes etc. are all good. My feeling is that a longer form more manual-style piece of documentation would still be of value (and we have something half written from memory), but that's not a good reason not to supplement it with other things.

ilyaZar commented:
Ok so the last changes are what has been asked for above, plus

1. I left the long form i.e. `arma::Col<unsigned int>`, not using a typedefs; but I added typedefs into the new discussion to do list #57 
2. I changed to `long` leaving a major revision of `int` vs `long` for later; also added this to the discussion To-Do list #57 
3. `history.h`: changed names form `SetAindices()` to `SetAIndices()` and from `GetAindices()` to `GetAIndices()` as I feel this also is more in line with our style of naming (but has not been asked for explicitly)

adamjohansen commented:
That all looks good to me.

adamjohansen reviewed (approved):
Sorry for holding things up, my comment in the main discussion that it all looked good was made after looking at the code I should probably have been a bit more formal about it.

adamjohansen commented:
The forced push seems to have confused the history of this discussion a bit but everything looks good to me and, by the look of it, @LeahPrice. Any objections, @eddelbuettel?

eddelbuettel reviewed (approved):
LGTM
